src/functional/histogram_compression.zig

```diff
@@ -0,0 +1,519 @@
+/// using "Piewice Constant Histogram" compression method and write the result to
+/// `compressed_values`. If an error occurs it is returned.
+pub fn compressPWCH(
+    uncompressed_values: []const f64,
+    compressed_values: *ArrayList(u8),
+    allocator: mem.Allocator,
+    error_bound: f32,
+) Error!void {
+    if (error_bound <= 0.0) {
+        return Error.UnsupportedErrorBound;
+    }
+
+    // The original implementation requires the maximum number of buckets which can be represented
+    // by a usize instead of `error_bound: f32`. Changing this requires modifications in
+    // `src/tersets.zig` and `src/capi.zig` files.
+    // TODO: Find the right of passing the maximum number of buckets.
+    const target_buckets: usize = @as(usize, @intFromFloat(@floor(error_bound)));
+    var histogram = try Histogram.init(allocator, target_buckets);
+    defer histogram.deinit();
+
+    for (uncompressed_values.items, 0..) |elem, index| {
+        try histogram.insert(elem, index);
+    }
+
+    for (0..2 * target_buckets) |index| {
+        const bucket: Bucket = histogram.getBucket(index);
+        try appendValueAndIndexToArrayList(bucket.computeError(), bucket.end, compressed_values);
+    }
+}
+
+/// Decompress `compressed_values` produced by "Piecewise Constant Histogram" and write the result
+/// to `decompressed_values`. If an error occurs it is returned. The implementation is similar to
+/// "Poor Man’s Compression" decompression function.
+pub fn decompress(
+    compressed_values: []const u8,
+    decompressed_values: *ArrayList(f64),
+) Error!void {
+    // The compressed representation is pairs containing a 64-bit float value and 64-bit end index.
+    if (compressed_values.len % 16 != 0) return Error.IncorrectInput;
+
+    const compressed_values_and_index = mem.bytesAsSlice(f64, compressed_values);
+
+    var compressed_index: usize = 0;
+    var uncompressed_index: usize = 0;
+    while (compressed_index < compressed_values_and_index.len) : (compressed_index += 2) {
+        const value = compressed_values_and_index[compressed_index];
+        const index: usize = @bitCast(compressed_values_and_index[compressed_index + 1]);
+        for (uncompressed_index..index) |_| {
+            try decompressed_values.append(value);
+        }
+        uncompressed_index = index;
+    }
+}
+
+/// Append `compressed_value` and `index` to `compressed_values`.
+fn appendValueAndIndexToArrayList(
+    compressed_value: f64,
+    index: usize,
+    compressed_values: *ArrayList(u8),
+) !void {
+    const value: f64 = @floatCast(compressed_value);
+    const valueAsBytes: [8]u8 = @bitCast(value);
+    try compressed_values.appendSlice(valueAsBytes[0..]);
+    const indexAsBytes: [8]u8 = @bitCast(index); // No -1 due to 0 indexing.
+    try compressed_values.appendSlice(indexAsBytes[0..]);
+}
+
+test "PWCH: HashedPriorityQueue with hashcontext for MergeError" {
```

Comment:
Each of the tests are complex, long, and tests multiple things, maybe split the tests into many much smaller tests and use private helper functions to not duplicate code. By having tests that each only check one thing, they become simpler to understand and thus check that they are correct. Also if a complex test with many asserts fails it is hard to understand why it failed while it is obvious for a test that only checks one thing.

Reply:
I understand why the tests look long. The testing is actually very short, most of the time is for creating the structures  and the expected result. I created an auxiliary function to generate a bounded random number but perhaps this is not enough to reduce the complexity of these tests.  I really tried to create other functions but I could not find any other overlapping code that can be moved to `tester.zig`.